### PR TITLE
core: Handle JSON as maps using JSX

### LINKED ISSUE #1716: Replace mochijson2 with jsx
Advantage: Erlang maps instead of hard-to-match `{struct, List}` structures.

### PR DESCRIPTION
**Pull request by ddeboer**
### Description

Maps are much easier to construct and unpack than the `{struct, …}` and `{array, …}` proplists that mochijson uses. Jjiffy, being a NIF, is arguably faster, but stability matters more. Though other libraries such as [RabbitMQ](https://github.com/rabbitmq/rabbitmq-common/issues/46) and [Erlasticsearch](https://github.com/tsloughter/erlastic_search) offer the choice between JSX and Jiffy, they usually have JSX as the default.

* Remove all calls to mochijson.
* Remove all calls to jiffy and the dependency.

Fix #1716.

### Checklist

- [x] documentation updated
- [ ] tests added
- [ ] no BC breaks

### CODE REVIEW

**Comment by mworrell** [comment]
👍 for not using a NIF based json library. As far as I can see so quickly, jiffy doesn't use dirty schedulers, so can block other execution.

**Comment by ddeboer** [comment]
And now green! Please review carefully, as I had to change around quite a lot.

**Comment by ddeboer** [comment]
Rebased to resolve merge conflicts.

File: apps/zotonic_mod_base/src/filters/filter_to_json.erl [hunk 1]
```diff
@@ -33,17 +33,15 @@
...
+    z_json:encode(Value).
 
 %% @doc Convert an Erlang list or tuple to JSON
 %% This function assumes that the all strings of the input term have the same
 %% character encoding. This encoding may be UTF-8 or ISO 8859-1 (also called Latin-1).
 %% @spec to_json(list() | tuple(), string(), #context{}) -> iodata()
 to_json(Value, "latin-1", _Context) ->
-    mochijson:encode(z_convert:to_json(Value));
-
+    z_json:encode(Value);
```
[the remark this answers is not in the document]
Comment (ddeboer) [reply apps/zotonic_mod_base/src/filters/filter_to_json.erl]: Yeah, removed that.

File: apps/zotonic_mod_facebook/src/models/m_facebook.erl [hunk 2]
```diff
@@ -116,15 +116,13 @@ do_graph_call(Method, Id, Connection, Args, Context)
...
     Request = make_httpc_request(Method, "https", "graph.facebook.com", Path, Query),
 
-    Payload = case httpc:request(Method, Request, [], []) of
+    case httpc:request(Method, Request, [], []) of
 		  {ok, {{_, 200, _}, _Headers, Body}} ->
-		      mochijson2:decode(Body);
+		      z_json:decode(Body);
 		  Other ->
 		      ?DEBUG({error, {http_error, element(1, Request), Other}}),
 		      []
```
[the remark this answers is not in the document]
Comment (ddeboer) [reply apps/zotonic_mod_facebook/src/models/m_facebook.erl]: Map and now processed as one, too. :wink:

[the remark this answers is not in the document]
Comment (mworrell) [reply apps/zotonic_mod_facebook/src/models/m_facebook.erl]: Should the error return of `[]` be a `#{}`

**Comment by ddeboer** [comment]
Now green and ready to merge.

File: apps/zotonic_core/src/support/z_json.erl [hunk 3]
```diff
@@ -19,12 +19,22 @@
...
+-export([
+    encode/1,
+    decode/1,
+    json_escape/1
 ]).
 
+%% @doc Encode an Erlang term to JSON.
+-spec encode(binary() | map()) -> binary().
+encode(Erlang) ->
+    jsx:encode(Erlang).
```
[the remark this answers is not in the document]
Comment (mworrell) [reply apps/zotonic_core/src/support/z_json.erl]: Maybe we should add a convenience function to map `{trans, ...}` tuples. Like `encode_trans`?
It needs a `z:context()` and will go recursively through the data structure, selecting the correct language versions.

[the remark this answers is not in the document]
Comment (mworrell) [reply apps/zotonic_core/src/support/z_json.erl]: Could also be part of the `z_trans` functions, might be better (separation of concerns and so on).

[the remark this answers is not in the document]
Comment (ddeboer) [reply apps/zotonic_core/src/support/z_json.erl]: Let’s do that in a separate PR, then.

[the remark this answers is not in the document]
Comment (mworrell) [reply apps/zotonic_core/src/support/z_json.erl]: Agree, then we can maybe also add a more generic mapper that also maps records to something that jsx can digest.

**Pull request merged**